libkbfs/chat_rpc.go

```diff
@@ -0,0 +1,296 @@
+//     * chat1.GetThreadLocal with pagination
+//     * No prev filled in on next pagination to go backward
+//   * How to reconcile renames, etc across channels?
+//     * It's hard to know if a long-ago writer updated a file, and
+//       later it was renamed by a prolific writer who made N more updates
+//       afterward.
+//     * For performance reasons, we probably just need to show the old
+//       update under the old file name.  Should be rare enough that
+//       it doesn't matter.
+
+// Getting real-time updates:
+//   * New kbfs-edits activity push notifications on notify-router
+//   * Mike will make ticket to auto-join non-chat channels,
+//     so they show up in `GetInboxAndUnboxLocal`.
+//   * Spot-edit the local edit history on each new push notification.
+
+// One layer over the service RPC connection that takes all needed
+// arguments (including topic type, etc), and passes it pretty
+// directly to the RPC.
+
+// Another, per-TLF layer to remember the resolved conversation ID and
+// send/read kbfs-edits messages.  It should also interface with the
+// local journal to show the unflushed journal data as part of the
+// updates.
+
+// Finally an inbox layer that can read the server inbox, and also
+// checks the journal status, to return the top set of conversations
+// at any given time.  Maybe it also subscribes to inbox notifications
+// of some kind.
+
+func membersTypeFromTlfType(tlfType tlf.Type) chat1.ConversationMembersType {
+	if tlfType == tlf.SingleTeam {
+		return chat1.ConversationMembersType_TEAM
+	}
+	return chat1.ConversationMembersType_IMPTEAMNATIVE
+}
+
+// GetConversationID implements the Chat interface.
+func (c *ChatRPC) GetConversationID(
+	ctx context.Context, tlfName tlf.CanonicalName, tlfType tlf.Type,
+	channelName string, chatType chat1.TopicType) (
+	chat1.ConversationID, error) {
+	// Try creating the conversation to get back the ID -- it's a
+	// no-op if the conversation already exists.
+	vis := keybase1.TLFVisibility_PRIVATE
+	if tlfType == tlf.Public {
+		vis = keybase1.TLFVisibility_PUBLIC
+	}
+
+	arg := chat1.NewConversationLocalArg{
+		TlfName:          string(tlfName),
+		TopicType:        chatType,
+		TlfVisibility:    vis,
+		MembersType:      membersTypeFromTlfType(tlfType),
+		IdentifyBehavior: keybase1.TLFIdentifyBehavior_CHAT_CLI,
+	}
+	res, err := c.client.NewConversationLocal(ctx, arg)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Conv.Info.Id, nil
+}
+
+// SendTextMessage implements the Chat interface.
+func (c *ChatRPC) SendTextMessage(
+	ctx context.Context, tlfName tlf.CanonicalName, tlfType tlf.Type,
+	convID chat1.ConversationID, body string) error {
+	arg := chat1.PostTextNonblockArg{
+		ConversationID:   convID,
+		TlfName:          string(tlfName),
+		TlfPublic:        tlfType == tlf.Public,
+		Body:             body,
+		IdentifyBehavior: keybase1.TLFIdentifyBehavior_CHAT_CLI,
+	}
+	_, err := c.client.PostTextNonblock(ctx, arg)
```

Comment:
The comment on the interface says "synchronously". Is either this method call or the comment not accurate?

Reply:
Good catch, the comment is wrong.  Mike recommended this RPC instead of the one I was originally planning and I forgot to update the comment.